Entities/AHuman.cpp

```diff
@@ -3502,34 +3502,36 @@ void AHuman::Update()
 			pThrown = dynamic_cast<ThrownDevice *>(m_pFGArm->GetHeldMO());
 			if (pThrown) {
 				if (m_Controller.IsState(WEAPON_FIRE)) {
-					if (m_ArmsState != THROWING_PREP/* || m_ThrowTmr.GetElapsedSimTimeMS() > m_ThrowPrepTime*/) {
+					if (m_ArmsState != THROWING_PREP) {
 						m_ThrowTmr.Reset();
-						if (!pThrown->ActivatesWhenReleased()) {
-							pThrown->Activate();
-						}
+						if (!pThrown->ActivatesWhenReleased()) { pThrown->Activate(); }
 					}
 					m_ArmsState = THROWING_PREP;
-					m_pFGArm->ReachToward(m_Pos + pThrown->GetStartThrowOffset().GetXFlipped(m_HFlipped));
+					m_pFGArm->ReachToward(m_Pos + (m_pFGArm->GetParentOffset() + pThrown->GetStartThrowOffset().RadRotate(m_AimAngle + m_AngularVel * g_TimerMan.GetDeltaTimeSecs())).GetXFlipped(m_HFlipped) * m_Rotation);
 				} else if (m_ArmsState == THROWING_PREP) {
 					m_ArmsState = THROWING_RELEASE;
-
-					m_pFGArm->SetHandPos(m_Pos + pThrown->GetEndThrowOffset().GetXFlipped(m_HFlipped));
+					//To-do: figure out how to properly use EndThrowOffset, since it doesn't play much a role for just one frame
+					m_pFGArm->SetHandPos(m_Pos + (m_pFGArm->GetParentOffset() + pThrown->GetEndThrowOffset().RadRotate(m_AimAngle * GetFlipFactor())).GetXFlipped(m_HFlipped) * m_Rotation);
 
 					MovableObject *pMO = m_pFGArm->ReleaseHeldMO();
 
-					if (pThrown->ActivatesWhenReleased()) {
-						pThrown->Activate();
-					}
 					if (pMO) {
-						pMO->SetPos(m_Pos + m_pFGArm->GetParentOffset().GetXFlipped(m_HFlipped) + Vector(m_HFlipped ? -15 : 15, -8));
+						pMO->SetPos(m_Pos + (m_pFGArm->GetParentOffset() + Vector(m_pFGArm->GetMaxLength(), -6.0F)).GetXFlipped(m_HFlipped).RadRotate(m_AimAngle * GetFlipFactor()) * m_Rotation);
 						float throwScalar = static_cast<float>(std::min(m_ThrowTmr.GetElapsedSimTimeMS(), static_cast<double>(m_ThrowPrepTime)) / m_ThrowPrepTime);
-						Vector tossVec(pThrown->GetMinThrowVel() + ((pThrown->GetMaxThrowVel() - pThrown->GetMinThrowVel()) * throwScalar), 0.5F * RandomNormalNum());
+						float maxThrowVel = pThrown->GetMaxThrowVel();
+						float minThrowVel = pThrown->GetMinThrowVel();
+						if (maxThrowVel == 0) {
+							// If throw velocity is decided by the arm and not by the device, then the mass of the device and angular velocity of the actor will be taken to account
+							maxThrowVel = (m_pFGArm->GetThrowStrength() + std::abs(m_AngularVel * 0.5F) * std::cos(rot)) / std::sqrt(std::abs(pMO->GetMass()) + 1.0F);
```

Comment:
I don't get why spinning around very fast means you throw better?

Also, why is rotAngle included here and cos'd? If I'm not getting this wrong, this means that if you're not standing straight up you'll throw weaker (cos(0) = 1, cos(45deg) = 0.7, etc.), and if you happen to throw while facing straight up you'll get no bonus angular vel bonus?
I guess rotangle will never be > 90deg cause of flipping, so no need to worry about throwing backwards when it's 180deg lol

Reply:
Yeah, that's what I was going for! On second thought, I don't know if it makes as much sense. What do you mean by "happen to throw while facing straight up"?

Reply:
cos(90deg) is 0. If your actor is destabilized but throwing, and their torso is facing upwards, that multiplier would be 0. I guess you've removed it though, so that's fine.

Does having angular vel affect this work out well?

Reply:
It feels alright, and the exponential (sqrt) factor makes sure that the boost isn't something insane. It did make more physical and/or visual sense without the clamp, but then again the kind of spring flail you can execute with units in CC is everything but realistic :p

Reply:
I think there's a miscommunication here???
The code is `maxThrowVel = (m_pFGArm->GetThrowStrength() + std::abs(m_AngularVel * 0.5F)) / std::sqrt(std::abs(pMO->GetMass()) + 1.0F);` so angularvel isn't being sqrted? Also, what clamp?

Reply:
Right, I meant the mass factor of the ThrownDevice... and now that you mention it, yeah the ang vel isn't clamped at all lol (no miscommunication, just me being a dummy)

In the end, if you find yourself spinning at 100 rad/s, just releasing the item should give it a fair nudge :p (but it's not like there's a situation where that would be very common)

Take for instance an edge case where an average actor (total radius = 25) is spinning at its maximum speed without filling TooFast conditions (2000 / 25 = 80), even then the angular vel boost is fairly minimal (80 / 2 = 40, which is ultimately diminished by the mass of the object)

There's obviously no realistic inertia calculation here but it's not practically unrealistic either.

Reply:
Oooh okay, yeah, I see the logic behind this now, and I'm cool with it.
Thanks for going through this with me, feel free to resolve :)